### PR TITLE
Alerting: rules delete API to check data source authorization

### PR DESCRIPTION
**Pull request by yuri-tceretian**
**What this PR does / why we need it**:
Currently, access to an alert rule (when fine-grained access control feature is enabled) is a composition of many permissions:
folder and data source(s), i.e. to add, update or delete a rule user must have permission to create/update/delete rules in a folder as well as query permission for all data sources used by the rule.

There are two delete API endpoints: one supports deletion of all rules from the folder, another - only rules that belong to the same group in the folder.

Both API methods operate with a group of rules, not individual rules. However, in general, a group can contain many rules, and in the worst-case scenario, a user that deletes a group may not have access to one or many rules in the group (because of the access to data sources). Those rules will not be even visible for that user. 

Therefore, the delete API should verify access to every rule that is being deleted, and do not delete that rule if the user does not have access to it.

Because of this behavior, I had to change the response code in the case when the group that user tries to delete does not exist. Currently, API responds with `404 Not Found`. This is changed to `202 Accepted`. This makes API consistent in the case when group exists but user does not have access some of alerts and when group does not exists at all

**Which issue(s) this PR fixes**:
Related: https://github.com/grafana/grafana/pull/46905

**Notes for reviewer**
- [x] tests

### CODE REVIEW

File: pkg/services/ngalert/api/api_ruler.go [hunk 1]
```diff
@@ -42,52 +42,86 @@ var (
...
+
+		if len(canDelete) == 0 {
+			return fmt.Errorf("%w to delete rules because user is not authorized to access data sources used by the rules", ErrAuthorization)
+		}
+
+		if len(cannotDelete) > 0 {
+			logger.Info("user cannot delete one or many alert rules because it does not have access to data sources. Those rules will be skipped", "expected", len(q.Result), "authorized", len(canDelete), "unauthorized", cannotDelete)
+		}
+
+		err = srv.store.DeleteAlertRulesByUID(ctx, c.SignedInUser.OrgId, canDelete...)
```
Comment (grobinson-grafana) [inline pkg/services/ngalert/api/api_ruler.go:101]:
```suggestion
		return srv.store.DeleteAlertRulesByUID(ctx, c.SignedInUser.OrgId, canDelete...)
```

File: pkg/services/ngalert/api/api_ruler.go [hunk 2]
```diff
@@ -42,52 +42,86 @@ var (
...
+	})
 
 	if err != nil {
-		if errors.Is(err, ngmodels.ErrRuleGroupNamespaceNotFound) {
-			return ErrResp(http.StatusNotFound, err, "failed to delete rule group")
+		if errors.Is(err, ErrAuthorization) {
+			return ErrResp(http.StatusUnauthorized, err, "")
 		}
-		return ErrResp(http.StatusInternalServerError, err, "failed to delete rule group")
+		return ErrResp(http.StatusInternalServerError, err, "failed to update rule group")
```
Comment (grobinson-grafana) [inline pkg/services/ngalert/api/api_ruler.go:112]: I think `delete` is better if this is an HTTP DELETE request?

**Comment by grobinson-grafana** [comment]
If a user attempts to delete all rules in the a folder, however does not have permission to delete one of the rules because of datasource permissions, following the delete request the user will see the folder as having no rules. Will the user then be able to delete the folder, because from their perspective all the rules in this folder have been deleted, or will the request fail because the folder still contains alert rules?

**Comment by yuri-tceretian** [comment]
Great question! The short answer is that a request to delete a non-empty folder should fail (not sure if that's the case currently. I will check)
When a user deletes all rules from the folder, and for example, it does not have access to one or many of them API deletes all that user have access to, and then the folder will not be visible in GET requests for that user.

**Review by grobinson-grafana — APPROVED** [review]

**Pull request merged**